group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TopicMetadata.java

```diff
@@ -106,16 +127,23 @@ public String toString() {
             "id=" + id +
             ", name=" + name +
             ", numPartitions=" + numPartitions +
+            ", partitionRackInfo=" + partitionRackInfo +
             ')';
     }
 
     public static TopicMetadata fromRecord(
         ConsumerGroupPartitionMetadataValue.TopicMetadata record
     ) {
+        // Converting the data type from a list stored in the record to a map.
+        Map<Integer, Set<String>> partitionRackInfo = new HashMap<>(record.partitionRackInfo().size());
+        for (ConsumerGroupPartitionMetadataValue.PartitionRackInfo info : record.partitionRackInfo()) {
+            partitionRackInfo.put(info.partition(), new HashSet<>(info.racks()));
```

Comment:
Let's make the set immutable here. You can use `Collections. unmodifiableSet` to do this.